docs/source/optimising.rst

```diff
@@ -0,0 +1,198 @@
+its logging infrastructure that Firedrake has extended by annotating many of
+its own functions with PETSc events. This allows users to easily generate
+informative flame graphs giving a lot of insight into the internals of
+Firedrake and PETSc.
+
+As an example, here is a flame graph showing the performance of the
+`scalar wave equation with higher-order mass lumping demo
+<https://firedrakeproject.org/demos/higher_order_mass_lumping.py.html>`_.
+It is interactive and you can zoom in on functions by clicking.
+
+.. raw:: html
+    :file: images/flame_graph_demo.svg
+
+One can immediately see that the dominant hotspots for this code are
+assembly and writing to output so any optimisation effort should be
+spent in those. Some time is also spent in ``firedrake.__init__`` but
+this corresponds to the amount of time spent importing Firedrake and
+would be amortized for longer-running problems.
+
+Flame graphs can also be generated for codes run in parallel with the
+reported times in the graph given by the maximum value across all ranks.
+
+Generating the flame graph
+~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+To generate a flame graph from your Firedrake script you need to:
+
+1. Run your code with the extra flag ``-log_view :foo.txt:ascii_flamegraph``.
+   For example:
+
+   .. code-block:: bash
+
+     $ python myscript.py -log_view :foo.txt:ascii_flamegraph
+
+   This will run your program as usual but output an additional file
+   called ``foo.txt`` containing the profiling information.
+
+2. Visualise the results. This can be done in one of two ways:
+    
+   * Generate an SVG file using the ``flamegraph.pl`` script from
+     `this repository <https://github.com/brendangregg/FlameGraph>`_
+     with the command:
+
+     .. code-block:: bash
+
+       $ ./flamegraph.pl foo.txt > foo.svg
+
+     You can then view ``foo.svg`` in your browser.
+
+   * Upload the file to `speedscope <https://www.speedscope.app/>`_ and view it there.
+
+Adding your own events
+~~~~~~~~~~~~~~~~~~~~~~
+
+It is very easy to add your own events to the flame graph and there
+are a few different ways of doing it. The simplest methods are:
+
+* With a context manager:
+
+  .. code-block:: python
+      
+      from firedrake.petsc import PETSc
+
+      with PETSc.Log.Event("foo"):
+          do_something_expensive()
+
+* With a decorator:
+
+  .. code-block:: python
+
+      from firedrake.petsc import PETSc
+
+      @PETSc.Log.EventDecorator("foo")
+      def do_something_expensive():
+          ...
+
+  If no arguments are passed to ``PETSc.Log.EventDecorator`` then the
+  event name will be the same as the function.
+
+Caveats
+~~~~~~~
+
+* The ``flamegraph.pl`` script assumes by default that the values
+  in the stack traces are sample counts. This means that if you
+  hover over functions in the SVG it will report the count in terms
+  of 'samples' rather than the correct unit of microseconds. A simple
+  fix to this is to include the command line option ``--countname us``
```

Comment:
Is flame graph.ph unicode safe? Could we use `-- countname µs`?

Reply:
No. Just tried and it did not work.